app/scripts/lib/mnemonic.ts

```diff
@@ -0,0 +1,418 @@
+/* eslint-disable no-bitwise, no-plusplus */
+
+/**
+ * @file Mnemonic utility for converting between BIP-39 mnemonics and wordlist
+ * indices without using strings in memory.
+ *
+ * This utility is designed to be used in environments where strings are not
+ * allowed in memory and where the process is long-running, such as the
+ * extension background process. It uses a compact trie to store the wordlist
+ * and convert between mnemonics and wordlist indices.
+ *
+ * The trie is built from the English wordlist, and can only be used with the
+ * English wordlist, as it assumes 26 lowercase letters and a resulting trie of
+ * limited depth.
+ *
+ * It has been optimized for memory size while maintaining reasonable search
+ * performance. It is typically 2x-5x faster than the string-based approach and
+ * uses only a little more memory (likely due to the application still importing
+ * the string-based wordlist elsewhere).
+ *
+ * This module is not suitable for using in the UI, as it has a non-trivial
+ * startup time.
+ *
+ * Additional startup improvements can be made by precomputing the trie and
+ * index. This would allow the trie to be used in the UI, but complicates
+ * maintenance.
+ */
+
+import { wordlist } from '@metamask/scure-bip39/dist/wordlists/english';
+
+/**
+ * 0xffff is the max uint16 value. we use it to represent a node that is not a
+ * terminal node, i.e., not the last character of a word.
+ */
+const NOT_TERMINAL = 0xffff;
+
+const ERROR_MESSAGE =
+  'Invalid mnemonic phrase: The mnemonic phrase contains an unknown word.';
+
+/**
+ * Bit-packing utilities for compact trie storage.
+ */
+
+/**
+ * Packs a 16-bit terminalIndex and 16-bit parentId into a Uint32.
+ *
+ * @param terminalIndex - 0xffff if not terminal, else (wordIndex)
+ * @param parentId - 0 if root, else node ID of parent
+ * @returns 32-bit packed value
+ */
+function packTermAndParent(terminalIndex: number, parentId: number): number {
+  return (parentId << 16) | terminalIndex;
+}
+
+/**
+ * Unpacks a Uint32 into terminalIndex.
+ *
+ * @param value - packed value
+ * @returns 16-bit terminalIndex
+ */
+function unpackTerminalIndex(value: number): number {
+  return value & 0xffff;
+}
+
+/**
+ * Unpacks a Uint32 into parentId.
+ *
+ * @param value - packed value
+ * @returns 16-bit parentId
+ */
+function unpackParentId(value: number): number {
+  return value >>> 16;
+}
+
+/**
+ * Packs a 6-bit letterFromParent and 26-bit childMask into a Uint32.
+ *
+ * @param letterFromParent - 0 if root, else (charCode + 1)
+ * @param childMask - 1 bit per child presence
+ * @returns 32-bit packed value
+ */
+function packLetterAndMask(
+  letterFromParent: number,
+  childMask: number,
+): number {
+  return (letterFromParent << 26) | (childMask & 0x03ffffff);
```

Comment:
This time we move the value we want on the left side to the right by 26 bits, leaving 26 bits to use as a "childMask".

Note: `0x03ffffff` is `00000011111111111111111111111111`.

We use `26` here because there are 26 possible letters in the english bip-39 wordlist. We use these bits (and some clever bitwise math) to determine if a trie node has children with a specific character. Neato!